### ISSUE TITLE
RocksDB ERROR when BPF is on

### ISSUE THREAD

**Issue by smgrol** [osquery/osquery#7884]
# Why RocksDB crash when BPF used?

When osqueryd work with BPF in the syslog I see errors:
1.
```
orbit[24125]: I1125 11:47:48.859995 24164 rocksdb.cpp:67] RocksDB: [ERROR] [table/block_based/block_based_table_reader.cc:1090] Encountered error while reading data from properties block IO error: While pread offset 4655206 len 40: /opt/orbit/osquery.db/004706.sst: Is a directory

orbit[24125]: I1125 11:47:48.955194 24164 rocksdb.cpp:67] RocksDB: [ERROR] [db/db_impl/db_impl_compaction_flush.cc:2624] Waiting after background flush error: IO error: While pread offset 4655206 len 40: /opt/orbit/osquery.db/004706.sst: Is a directoryAccumulated background error counts: 1

orbit[24125]: E1125 11:49:16.361975 25035 shutdown.cpp:79] Error adding new results to database for query pack/Global/bpf_socket_events (test)-1: IOError: Is a directory
```
2.
```
orbit[13777]: I1125 11:43:09.381443 13824 rocksdb.cpp:67] RocksDB: [WARN] [db/error_handler.cc:387] Background IO error IO error: While appending to file: /opt/orbit/osquery.db/005943.sst: Bad file descriptor

orbit[13777]: I1125 11:43:09.381582 13824 rocksdb.cpp:67] RocksDB: [ERROR] [db/db_impl/db_impl_compaction_flush.cc:2624] Waiting after background flush error: IO error: While appending to file: /opt/orbit/osquery.db/005943.sst: Bad file descriptorAccumulated background error counts: 1

orbit[13777]: E1125 11:44:52.375362 19350 shutdown.cpp:79] Error adding new results to database for query pack/Global/ebpf proccess events (test): IOError: Bad file descriptor
```

After that agent is restarted.

**Why this errors occurs? Maybe I can tune some flags for exclude this error?**
It was not possible to find the reason why this error occurs. The test machines are not heavily loaded. An agent can restart about 10 times per day.

**Used BPF queries:**
1.
`SELECT uid, gid, local_address, local_port, path, pid, remote_address, remote_port, ntime, duration, family, exit_code FROM bpf_socket_events WHERE path NOT IN ('/usr/bin/hostname') AND remote_address NOT IN ('127.0.0.1', '169.254.169.254', '', '0000:0000:0000:0000:0000:0000:0000:0001', '::1', '0000:0000:0000:0000:0000:ffff:7f00:0001', 'unknown', '0.0.0.0', '0000:0000:0000:0000:0000:0000:0000:0000');`
2.
`SELECT cmdline, ntime, cwd, gid, uid, parent, path, pid FROM bpf_process_events;
`
**Flags used:**
```
  table_delay: 200
  augeas_lenses: /usr/share/augeas/lenses/dist
  logger_plugin: tls
  disable_events: false
  watchdog_delay: 180
  events_expiry: 120
  buffered_log_max: 5000000
  disable_watchdog: false
  enable_bpf_events: true
  logger_tls_period: 1
  logger_tls_max_lines: 4000
  watchdog_memory_limit: 2048
  bpf_buffer_storage_size: 2048
```
On the systemd unit we set CPUQuota=10%.

Osquery version: 5.6.0
OS: Debian 11.
VMs in test environment has above 8 GB RAM, 4 CPU.

**Comment by Smjert** [comment]
Hello @smgrol thanks for the bug report; I suspect that the RocksDB errors and state are not directly caused by BPF table, because that uses the same interface towards RocksDB as any other.

I suspect there might be a secondary problem, likely due to the CPUQuota, the memory usage and the watchdog.
So first of all a CPUQuota of 10% (and even the default we have of 20%) can be ok, or be pretty low, depending on how many cores the machine is runs on has, because that % is over a single core.

We have observed in the past that some code multithreaded code in osquery which buffers events, but doesn't put a limit to the buffering, when having low CPU available it may start increasing the memory usage indefinitely, due to the processing side being always slower than the collecting side.
This then causes the watchdog to kill osquery, and if this happens often it may corrupt the RocksDB database.

We will check this theory, but in the meanwhile if you still have the logs around, I would check in `journalctl` if there are messages like this from osqueryd:
`Memory limits exceeded: <amount>`

**Comment by smgrol** [comment]
hello @Smjert 

1. when RocksDB fail we don't get errors like `Memory limits exceeded: <amount>`
2. if i don't change default CPUQuota param it systemd unit and don't use watchdog limits, osquery agent work good and don't restarted 
<img width="553" alt="image" src="https://user-images.githubusercontent.com/70931518/207780305-e724889d-4a46-4b8b-8845-f1d7352e12e3.png">
<img width="553" alt="image" src="https://user-images.githubusercontent.com/70931518/207780345-ce2f1cf4-c801-476f-83fd-25c58c604676.png">
ps: on tested host in the previous test agent restarted over 5-7 times per night.        


**questions**:
1) so what can we change in the work of the agent so that it works stably and can work with a limit of up to 10-15% CPU? 20% may be too much for some nodes.

2) if we reduce the number of scheduled tasks and the data received from them, can we minimize the number of agent crashes?

**Comment by Smjert** [comment]
@smgrol Just to clarify, those memory limit exceeded will likely happen before the RocksDB error appears.
Are you 100% sure that there are no messages like that in all the logs for that day in `journalctl` or syslog?

Because if it's not the watchdog, then it means osquery is actually crashing, just for having low CPU usage, apparently.
If it's a crash, it would really be helpful to have a stack trace. I should've asked if you have core dumps enabled and `coredumpctl` handling them, by checking `coredumpctl list`.
If you can see entries for osqueryd, then we can get a stack trace. If so and it's possible to install additional packages on one of those machines having the issue:

1. Install the debug info package, either via `apt` if you have the repositories with `sudo apt install osquery-dbgsym`, or downloading them from there, selecting `Debug` in the `Release Type` on top https://osquery.io/downloads/debug/5.6.0, the installing them with `sudo apt install ./<packagename>`.
2. Install gdb via apt
3. If `coredumpctl list` shows entries for osqueryd with the word `present`, then the core dump is already there, please go to the next point, otherwise restore the problematic configuration and have osquery run until it gets restarted.
4. Check with `coredumpctl list` if you can see a new entry for osqueryd which should have the word `present`, meaning that the coredump is there.
5. Move yourself into a folder you can store a log, then run `coredumpctl debug osqueryd`, this should open gdb and give on the top a preliminary stack trace and reason; I would ignore it and execute the following commands:
```
set logging on
thread apply all bt
```
Then press `c` and enter so that it automatically goes through the whole stack trace. Then you can quit it with `q` and enter.
In the folder you are, you should have a `gdb.txt` file, that will contain the stacktrace, please upload it here.

If you could do the above it would really appreciated!

As for your questions:
1. If you mean 10-15% of the **whole** CPU (all the cores), then the CPUQuota cannot stay fixed unless all the machines have the same number of cores. Again you have to consider that CPUQuota % is on a single core, so if you have a 4 cores, 100% of the whole CPU is actually 400% for the CPUQuota. So you have to take the whole CPU % you want to keep the process at and then multiply that for the number of cores you have in the system.
If 10%, that will give you 40% which has to be put in the CPUQuota.
Keep also in mind that the watchdog CPU utilization limit is by default 10% of the whole CPU, so if you increase the CPUQuota beyond that you also have to increase the watchdog one with the flag `watchdog_utilization_limit`.

2. It might be that actually giving higher limits as above it will reduce or eliminate the crashing. The osquery workload has to be normally tuned for the machine.
But if osquery was crashing then there's also a bug that we have to fix.

**Comment by smgrol** [comment]
@Smjert 
yes, i am sure, this is log to ~10 min before agent is restarted
```
Dec 15 15:20:42 <hostname> orbit[2903946]: 2022-12-15T15:20:42+03:00 INF calling flags update
Dec 15 15:21:13 <hostname> orbit[2903946]: 2022-12-15T15:21:13+03:00 INF calling flags update
Dec 15 15:21:43 <hostname> orbit[2903946]: 2022-12-15T15:21:43+03:00 INF calling flags update
Dec 15 15:22:13 <hostname> orbit[2903946]: 2022-12-15T15:22:13+03:00 INF calling flags update
Dec 15 15:22:43 <hostname> orbit[2903946]: 2022-12-15T15:22:43+03:00 INF calling flags update
Dec 15 15:23:13 <hostname> orbit[2903946]: 2022-12-15T15:23:13+03:00 INF calling flags update
Dec 15 15:23:43 <hostname> orbit[2903946]: 2022-12-15T15:23:43+03:00 INF calling flags update
Dec 15 15:24:13 <hostname> orbit[2903946]: 2022-12-15T15:24:13+03:00 INF calling flags update
Dec 15 15:24:43 <hostname> orbit[2903946]: 2022-12-15T15:24:43+03:00 INF calling flags update
Dec 15 15:25:14 <hostname> orbit[2903946]: 2022-12-15T15:25:14+03:00 INF calling flags update
Dec 15 15:25:44 <hostname> orbit[2903946]: 2022-12-15T15:25:44+03:00 INF calling flags update
Dec 15 15:26:14 <hostname> orbit[2903946]: 2022-12-15T15:26:14+03:00 INF calling flags update
Dec 15 15:26:44 <hostname> orbit[2903946]: 2022-12-15T15:26:44+03:00 INF calling flags update
Dec 15 15:27:15 <hostname> orbit[2903946]: 2022-12-15T15:27:15+03:00 INF calling flags update
Dec 15 15:27:45 <hostname> orbit[2903946]: 2022-12-15T15:27:45+03:00 INF calling flags update
Dec 15 15:28:15 <hostname> orbit[2903946]: 2022-12-15T15:28:15+03:00 INF calling flags update
Dec 15 15:28:45 <hostname> orbit[2903946]: 2022-12-15T15:28:45+03:00 INF calling flags update
Dec 15 15:29:15 <hostname> orbit[2903946]: 2022-12-15T15:29:15+03:00 INF calling flags update
Dec 15 15:29:45 <hostname> orbit[2903946]: 2022-12-15T15:29:45+03:00 INF calling flags update
Dec 15 15:30:15 <hostname> orbit[2903946]: 2022-12-15T15:30:15+03:00 INF calling flags update
Dec 15 15:30:26 <hostname> orbit[2904276]: E1215 15:30:26.873639 2905022 certificates.cpp:260] osquery::tables::(anonymous namespace)::OpenSSLError[0] (Failed to open the following file: "/etc/pki/tls/certs/ca-bundle.crt")
Dec 15 15:30:46 <hostname> orbit[2903946]: 2022-12-15T15:30:46+03:00 INF calling flags update
Dec 15 15:31:16 <hostname> orbit[2903946]: 2022-12-15T15:31:16+03:00 INF calling flags update
Dec 15 15:31:46 <hostname> orbit[2903946]: 2022-12-15T15:31:46+03:00 INF calling flags update
Dec 15 15:32:16 <hostname> orbit[2903946]: 2022-12-15T15:32:16+03:00 INF calling flags update
Dec 15 15:32:44 <hostname> orbit[2904276]: I1215 15:32:44.472404 2905022 rocksdb.cpp:67] RocksDB: [WARN] [db/error_handler.cc:387] Background IO error IO error: While appending to file: /opt/orbit/osquery.db/004276.log: Bad file descriptor
Dec 15 15:32:44 <hostname> orbit[2904276]: E1215 15:32:44.671681 2905022 shutdown.cpp:79] Error logging the results of query: pack/Global/bpf_process_events-1: IOError: Bad file descriptor
Dec 15 15:32:56 <hostname> orbit[2903946]: 2022-12-15T15:32:56+03:00 ERR unexpected exit error="deregistering extension: write unix @->/opt/orbit/orbit-osquery.em: write: broken pipe"
Dec 15 15:33:02 <hostname> systemd[1]: orbit.service: Succeeded.
Dec 15 15:33:02 <hostname> systemd[1]: orbit.service: Consumed 3min 53.000s CPU time.
Dec 15 15:33:03 <hostname> systemd[1]: orbit.service: Scheduled restart job, restart counter is at 1.
```
this grep by 'Memory limits'
<img width="325" alt="Untitled" src="https://user-images.githubusercontent.com/70931518/207895965-f38ed402-34d6-4b47-99cc-9f5db84a783a.png">
<img width="400" alt="www" src="https://user-images.githubusercontent.com/70931518/207896320-731e4cae-3efc-43d9-898e-2d62bf75a87c.png">


how i can get stack trace for osqueryd if osqueryd started by orbit?
`coredumpctl list` return `No coredumps found.`

ps: thanks for the detailed explanation of CPUQuota, this is my mistake in understanding this parameter

**Comment by smgrol** [comment]
@Smjert 
help me please with full understanding this:
1. if I set `CPUQuota=160%` for **16** **cores** (which is mean 10% per core, right?), i need change `watchdog_utilization_limit` or not? 
2. if yes, which values i can set?

**Comment by Smjert** [comment]
1. It more means 10% of total CPU usage; pratically you can still have 2 cores, one used 100%, the other 60% and the rest 0 (at least from osqueryd), or any other kind of distribution that sums up to 160%, depending on which thread is causing work in osquery and the scheduling of the OS.

2. By default `watchdog_utilization_limit` is set at 10% total cpu usage. It depends on you on what you want to set it, but I would say that it should be slightly higher than what you set the CPUQuota (divided by the number of cores).
In theory this check shouldn't be needed because the CPUQuota hard limits the CPU usage, but it might be considered as a safety measure in the case something wrong happens with the CPUQuota.

**Comment by smgrol** [comment]
@Smjert thanks for helping me understand

if it need https://github.com/osquery/osquery/issues/7884#issuecomment-1353252246 i can get stack trace for osqueryd

**Comment by Smjert** [comment]
> @Smjert thanks for helping me understand
> 
> if it need [#7884 (comment)](https://github.com/osquery/osquery/issues/7884#issuecomment-1353252246) i can get stack trace for osqueryd

It would be very useful to get a stack trace; it seems you might have core dumps disabled but I cannot closely guide you on how to re-enable them, because there are several ways they might have been disabled.
Some information can be seen here, in the Configuration section: https://manpages.debian.org/testing/systemd-coredump/systemd-coredump.8.en.html.

At the same time something may have been changed in `/etc/security/limits.conf` or a file in `/etc/security/limits.d/`

**Comment by smgrol** [comment]
@Smjert hello! i send this file in the slack

**Comment by ZetaTwo** [comment]
We have observed similar FD related issues in osquery on heavily loaded servers. An additional symptom we also saw was the osquery extension socket randomly being closed causing issues.

We have tried running a version of osquery built with https://github.com/facebook/rocksdb/pull/14104 and the issues now seems gone. We would previously see errors within ~20 minutes on one affected host and have now run for many hours without a single issue.

Some of the symptoms we observed are very similar to what's described in this issue https://github.com/osquery/osquery/issues/7954 so this might be the fix for that one too.

**Comment by ZetaTwo** [comment]
Unfortunately, it turns out I spoke prematurely. The issue is not gone despite this change. We are now running a valgrind build of osquery and seeing 1000s of fd double close errors. I will report back once we have something more to go on

**Comment by nickrmc83** [comment]
@ZetaTwo did you make any progress with the above?

**Comment by ZetaTwo** [comment]
I spoke a bit with @sharvilshah and he mentioned that the whole BPF part is getting a huge overhaul (https://github.com/osquery/osquery/pull/8743) so instead of trying to fight with these issues which seem to be related (or at least exaggerated by using BPF), we tried to switching to audit events instead. This seems to have completely gotten rid of the issues, even in high load environment (at least the few test hosts we looked at) and we are now rolling out this version broader.

We might revisit BPF after the overhaul.

In short: we gave up on this and switched to audit events instead of BPF for the time being
